File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {boolean[]} hasApple
 * @return {number}
 */
var minTime = function(n, edges, hasApple) {
  const grid = Array.from({ length: n }, () => []);
  
  const dfs = (curr, prev) => {
    let ans = 0;
    
    for (let next of grid[curr]) {
      if (next === prev) continue;
      
      ans += dfs(next, curr);
    }
    
    return ans || hasApple[curr] ? ans + 2 : ans;
  }
  
  edges.forEach(([a, b]) => {
    grid[a].push(b);
    grid[b].push(a);
  })
  
  return Math.max(dfs(0, -1) - 2, 0);
};

File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        def dfs(curr, prev):
            ans = sum(dfs(next, curr) for next in grid[curr] if next != prev)
            if ans or hasApple[curr]:
                ans += 2
            return ans
          
          
        grid = [[] for _ in range(n)]
        
        for u, v in edges:
            grid[u].append(v)
            grid[v].append(u)
        
        return max(dfs(0, -1) - 2, 0)
            